hack/tools/logcheck/README.md

```diff
@@ -1,5 +1,69 @@
-This directory contains tool for checking use of unstructured logs in a package. It is created to prevent regression after packages have been migrated to use structured logs.
+This directory contains a linter for checking log calls. It was originally
+created to detect when unstructured logging calls like `klog.Infof` get added
+to files that should only use structured logging calls like `klog.InfoS`
+and now also supports other checks.
 
-**Installation:**`go install k8s.io/klog/hack/tools/logcheck`
-**Usage:** `$logcheck.go <package_name>`
+# Installation
+
+`go install k8s.io/klog/hack/tools/logcheck`
+
+# Usage
+
+`$logcheck.go <package_name>`
 `e.g $logcheck ./pkg/kubelet/lifecycle/`
+
+# Configuration
```

Comment:
I don't find this configuration format better than `hack/verify-structured-logging.sh`. Have we first looked if it is possible to simplify the script before writing our own custom configuration format?

Reply:
Did you look at the [commit message](https://github.com/kubernetes/klog/pull/297/commits/99509e72aa61400ea26d29aa3c5c3e435c330664)? It cites several other advantages of integrating with golangci-lint:
- faster checking because finding files and parsing is shared
  with other linters
- support for // nolint:logcheck

Moving the logic of what to check where out of the script and into the checker is required for that integration.

This new configuration format also scales better: instead of maintaining different lists of files where unstructured calls are banned and another list where all klog logging calls are banned we only need to maintain one config file.

Reply:
Regarding the script: others are welcome to try and simplify it, but I consider that a dead end and won't spend time on it. The issue about fixing it has been open for months without progress, too.

Reply:
Not sure if the advantages of golangci-lint you listed impact k8s enough to motivate this change. Personally I would really want to avoid introducing custom configuration as it generates much more work long term. To ensure maintainability of kubernetes we need to ensure that we only consider cost of implementation, but also how much work other contributors will spend time to understand, debug, fix and extend this format.

I would prefer for introduction of contextual logging to logcheck not to include this. However I don't have enough experience in maintaining validation scripts for K8s, so maybe I'm wrong. I think best would be to get some feedback from `./hack/` directory owners as they will be de final approvers on the configuration.
/cc bentheelder cblecker deads2k fejta lavalamp liggitt pwittrock SataQiu spiffxp sttts mikedanese

Reply:
/cc @bentheelder @cblecker @deads2k @fejta @lavalamp @liggitt @pwittrock @SataQiu @spiffxp @sttts @mikedanese

Guys, can anyone comment on this question: do you prefer to integrate log checking into the common golanci-lint invocation (drawback: a more complex configuration file is needed) or maintaining several different scripts under hack, one for structured logging and another for contextual logging (drawbacks: current script is complex and broken, slower overall, no support for nolint:logcheck).

The support for integration with golanci-lint is in this PR. The prototype integration into Kubernetes is currently in https://github.com/kubernetes/kubernetes/compare/master...pohly:logcheck

Reply:
Let me add that I had asked about this before going in this direction: https://kubernetes.slack.com/archives/C09QZ4DQB/p1643049767036400